idle_sweep_test.go

```diff
@@ -0,0 +1,239 @@
+		SetIdleCheckInterval(30 * time.Second).
+		SetMaxIdleTime(3 * time.Minute).
+		SetTimeNow(clock.Now).
+		NoRelay()
+
+	clientOpts := testutils.NewOpts().
+		SetTimeNow(clock.Now)
+
+	testutils.WithTestServer(t, serverOpts, func(ts *testutils.TestServer) {
+		ts.RegisterFunc("echo", func(ctx context.Context, args *raw.Args) (*raw.Res, error) {
+			return &raw.Res{}, nil
+		})
+
+		client := ts.NewClient(clientOpts)
+		raw.Call(ctx, client, ts.HostPort(), ts.ServiceName(), "echo", nil, nil)
+
+		// Both server and client now have an active connection. After 3 minutes they
+		// should be cleared out by the idle sweep.
+		for i := 0; i < 2; i++ {
+			clock.Elapse(1 * time.Minute)
+
+			serverTicker.Tick()
+			time.Sleep(testutils.Timeout(10 * time.Millisecond))
+
+			assert.Equal(t, 1, numConnections(ts.Server()))
+			assert.Equal(t, 1, numConnections(client))
+		}
+
+		clock.Elapse(90 * time.Second)
+		serverTicker.Tick()
+
+		// Allow some time for the disconnect to propagate
+		time.Sleep(testutils.Timeout(10 * time.Millisecond))
+
+		assert.Equal(t, 0, numConnections(ts.Server()))
+		assert.Equal(t, 0, numConnections(client))
+	})
+}
+
+// Validates that outbound idle connections are dropped.
+func TestClientBasedSweep(t *testing.T) {
+	ctx, cancel := NewContext(time.Second)
+	defer cancel()
+
+	clientTicker := testutils.NewFakeTicker()
+	clock := testutils.NewStubClock(time.Now())
+
+	serverOpts := testutils.NewOpts().
+		SetTimeNow(clock.Now).
+		NoRelay()
+
+	clientOpts := testutils.NewOpts().
+		SetTimeNow(clock.Now).
+		SetTimeTicker(clientTicker.New).
+		SetMaxIdleTime(3 * time.Minute).
+		SetIdleCheckInterval(30 * time.Second)
+
+	testutils.WithTestServer(t, serverOpts, func(ts *testutils.TestServer) {
+		ts.RegisterFunc("echo", func(ctx context.Context, args *raw.Args) (*raw.Res, error) {
+			return &raw.Res{}, nil
+		})
+
+		client := ts.NewClient(clientOpts)
+		raw.Call(ctx, client, ts.HostPort(), ts.ServiceName(), "echo", nil, nil)
+
+		// Both server and client now have an active connection. After 3 minutes they
+		// should be cleared out by the idle sweep.
+		clientTicker.Tick()
+		time.Sleep(testutils.Timeout(10 * time.Millisecond))
+
+		assert.Equal(t, 1, numConnections(ts.Server()))
+		assert.Equal(t, 1, numConnections(client))
+
+		clock.Elapse(180 * time.Second)
+		clientTicker.Tick()
+
+		// Allow some time for the disconnect to propagate
+		time.Sleep(testutils.Timeout(10 * time.Millisecond))
+
+		assert.Equal(t, 0, numConnections(ts.Server()))
+		assert.Equal(t, 0, numConnections(client))
+	})
+}
+
+// Validates that a relay also disconnects idle connections - both inbound and
+// outbound.
+func TestRelayBasedSweep(t *testing.T) {
+	ctx, cancel := NewContext(time.Second)
+	defer cancel()
+
+	relayTicker := testutils.NewFakeTicker()
+	clock := testutils.NewStubClock(time.Now())
+
+	serverOpts := testutils.NewOpts().
+		SetTimeNow(clock.Now).
+		SetRelayOptionsFn(func(relayOpts *testutils.ChannelOpts) {
```

Comment:
do we need the relay options fn? can we just create a new server which doesn't have tehse options by using `ts.NewServer`, and ignore `ts.Server`?

the `testutils` channel logic is pretty complicated, so I'd like to avoid extra logic there if possible.

Reply:
Done